### DISCUSSION TITLE
Where should I add the code for a custom directive?

### DISCUSSION THREAD

**Discussion by vmgustavo** [marp-team/marp#319] [Q&A] [answered]
It is not clear how to add the custom directive code in this page https://marpit.marp.app/directives?id=custom-directives
How should I do it? Is there any other page that explains it?

**Comment by yhatt** [comment, marked as answer]
> Marpit instance has [`customDirectives.global` and `customDirectives.local` object](https://marpit-api.marp.app/marpit#customDirectives) to allow adding directives as you like.

The example code is assuming to have `marpit` instance. This is a full example:

```javascript
import { Marpit } from '@marp-team/marpit'

const marpit = new Marpit()

// Define $theme global directive for aliasing to theme directive
marpit.customDirectives.global.$theme = (value) => {
  return { theme: value }
}

// Define colorPreset local directive for setting multiple local directives at once
marpit.customDirectives.local.colorPreset = (value) => {
  switch (value) {
    case 'sunset':
      return { backgroundColor: '#e62e00', color: '#fffff2' }
    case 'dark':
      return { backgroundColor: '#303033', color: '#f8f8ff' }
    default:
      return {}
  }
}

// Render Markdown, and use the result as you like :)
const { html, css } = marpit.render(`
<!-- $theme: foobar -->
<!-- colorPreset: sunset -->

...
`)
```

**Reply by vmgustavo** [reply]
> The example code is assuming to have marpit instance

I am quite new to this, what is a marpit instance?

**Reply by yhatt** [reply]
If you had followed [the usage of Marpit](https://marpit.marp.app/usage), you already should have made an instance of Marpit class.

**Comment by OlegHahm** [comment]
I'm struggling to understand this example. I'm running marp-cli to create HTML slides from my markdown file. My understanding so far is, that marp-cli will by default use the Marp-Core as engine. Is there a way to extend this engine by custom directives?

**Reply by yhatt** [reply]
[Marp Core class](https://github.com/marp-team/marp-core/blob/2e3509905d14112f44e4a4c8d9e8913b6fee0862/src/marp.ts#L34) is [an inherited class](https://en.wikipedia.org/wiki/Inheritance_(object-oriented_programming)) from Marpit class, so _Marp Core instance is also a Marpit instance_ ([See also the architecture diagram in the project contributing guideline](https://github.com/marp-team/.github/blob/master/CONTRIBUTING.md#architecture)). Thus, a way to define custom directives is exactly same as [Marpit](https://marpit.marp.app/directives?id=custom-directives).

[As described at Marp CLI documentation](https://github.com/marp-team/marp-cli#spec), the custom engine for Marp CLI should return an instance of Marpit (= also allowed Marp Core instance) created by the constructor option passed as the first argument of the function.

```javascript
// engine.mjs
import { Marp } from '@marp-team/marp-core'

export default (constructorOptions) => {
  // Create Marp instance, from options passed as the first argument of the functional engine
  const marp = new Marp(constructorOptions);

  // Define custom directives to the created instance
  marp.customDirectives.local.colorPreset = (value) => {
    // ...
  };

  // And return a modified instance
  return marp;
}
```

> [!NOTE]
> For just extending Marp CLI's built-in Marp Core, you don't need create an instance from scratch. You can pick up a prepared Marp instance through `marp` getter.
>
> ```javascript
> export default ({ marp }) => {
>   marp.customDirectives.local.colorPreset = (value) => {
>     // ...
>   };
> 
>   return marp;
> }
> ```

**Reply by OlegHahm** [reply]
Great, thanks a bunch!
